report/5.Solution/StructureCodebase.tex

```diff
@@ -0,0 +1,113 @@
+\section{Structure of Codebase}
+The codebase is twofold, where one part is for object detection and the other part is for movement and interaction with the motors.
+\todo{figure out whether the python part also does the calibration or not}
+The object detection part, which is run on a Raspberry Pi, is written in Python 3{.}7, and the movement elements are run on the NXT written i C, version C90 with GNU-extensions.
+
+
+\subsection{Object Detection}
+The codebase is seperated into modules with the intent of creating a system where each module can be exchanged based on different contexts and further development.
+The following modules exist:
+\begin{itemize}
+	\item Object Localization Algorithm
+	\item Output / Communication
+	\item Input / Webcam
+\end{itemize}
+
+These are all utilized by a controller class, called \texttt{FlatController}, that ties the solution together, which has the signature shown in snippet~\ref{lst:FlatControllerInit}.
+\begin{lstlisting}[language=Python,label={lst:FlatControllerInit},caption={Initialization method of the \texttt{FlatController} class}]
+def __init__(self,
+	algorithm: Callable[[np.ndarray], Vector],
+	video_controller: webcam.VideoController,
+	output_devices: Union[OutputDevice, List[OutputDevice], None] = None,
+	) -> None:
+	"""
+	Initializes the controller
+	:param output_devices: The device to send data to
+	:param algorithm: The algorithm to use for image processing
+	:param video_controller: The controller handling the input video
+	"""
+\end{lstlisting}
+
+The order of the parameters doesn't correspond to the order of relevance, but rather is a requirement as two of the parameters have default arguments.
+This method takes, as the signature shows, a function, \texttt{algorithm} with an array as input and a vector as output.
+This array is the array of pixels in a given frame.
+The intent of this algorithm is to take a given frame, and localize the target within the frame.
+
+
+\todo{implement the video output debug to actually be what i've written below}
+The \texttt{output\_devices} parameter is a class that handles transmitting the location data to the NXT, but this can be overwritten to use a video feed as output as well, with the intent of debugging. 
+This argument can either be a singular \texttt{OutputDevice} or multiple, which is relevant when the user both wants to debug but also transmit the information to the NXT device.
+
+
+\todo{maybe change the capture\_type to also be either a class or a function, so we follow the same paradigm of coding all the way through}
+Lastly the \texttt{capture\_type} parameter handles from which device the video feed should come from.
+By default this should be a camera feed, but it might be relevant to use a specific video file in some contexts, for instance when doing tests.
```

Comment:
```suggestion
By default, this should be a camera feed, but it might be relevant to use a specific video file in some contexts, for instance when doing tests.
```